micro-rdk/src/common/data_manager.rs

```diff
@@ -27,6 +27,23 @@ use thiserror::Error;
 // the smaller amount of available RAM, we've halved it
 static MAX_SENSOR_CONTENTS_SIZE: usize = 32000;
 
+/// Allow for a C project using micro-RDK as a library to implement a callback to be run
+/// whenever data has successfully been uploaded. The callback should be identical in signature
+/// to the weak symbol declared here
+///
+/// Ex. for ESP-IDF project
+/// void data_upload_callback(void) {
+///     ESP_LOGI("proj_main", "hit data upload callback");
+/// }
+///
+/// # Safety
+/// The implemented callback should be simple. Long-running operations
+/// would endanger the stability of the data management feature
+#[cfg(feature = "esp32")]
+#[linkage = "weak"]
+#[no_mangle]
+pub unsafe extern "C" fn data_upload_callback() {}
```

Comment:
All quibbles but:

- I'd call this a "hook" not a "callback".
- The name right now doesn't tell me exactly when the call is made: before data is uploaded or after?
- I also think the function should include `micro-rdk` in the name to keep it namespaced, and `data_manager` to connect it to the struct that uses it.

So, maybe `micro_rdk_data_manager_post_upload_hook`?